File: design-pattern-go/src/patterns/creational/builder/CarBuilder.go
package builder

type CarBuilder struct {
	car car
}

func NewCarBuilder() *CarBuilder {
	return &CarBuilder{car{
		Doors:      2,
		Motor:      "4L",
		Lts:        1.4,
		AC:         false,
		Electronic: false,
	}}
}
func (b *CarBuilder) SetDoors(doors int) *CarBuilder {
	b.car.Doors = doors
	return b
}
func (b *CarBuilder) SetMotors(motor string) *CarBuilder {
	b.car.Motor = motor
	return b
}
func (b *CarBuilder) SetLts(lts float32) *CarBuilder {
	b.car.Lts = lts
	return b
}
func (b *CarBuilder) SetElectronic(electronic bool) *CarBuilder {
	b.car.Electronic = electronic
	return b
}
func (b *CarBuilder) Build() car {
	return b.car
}

File: design-pattern-ts/src/patterns/creator/builder/CarBuilder.ts
import { Car } from "./Car"

export class CarBuilder {
	private car: Car
	static create(): CarBuilder {
		return new CarBuilder()
	}
	private constructor() {
		this.car = new Car(2, "4L", 1.4, false, false)
	}
	setDoors(doors: number): CarBuilder {
		this.car.doors = doors
		return this
	}
	setMotor(motor: string): CarBuilder {
		this.car.motor = motor
		return this
	}
	setLts(lts: number): CarBuilder {
		this.car.lts = lts
		return this
	}
	setAc(ac: boolean): CarBuilder {
		this.car.ac = ac
		return this
	}
	setElectronics(electronics: boolean): CarBuilder {
		this.car.electronics = electronics
		return this
	}
	build(): Car {
		return this.car
	}
}
